x/axelarnet/keeper/coin.go

```diff
@@ -0,0 +1,155 @@
+package keeper
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ibctypes "github.com/cosmos/ibc-go/v2/modules/apps/transfer/types"
+
+	"github.com/axelarnetwork/axelar-core/x/axelarnet/exported"
+	"github.com/axelarnetwork/axelar-core/x/axelarnet/types"
+	"github.com/axelarnetwork/utils/funcs"
+)
+
+type coin struct {
+	sdk.Coin
+	ctx      sdk.Context
+	k        Keeper
+	nexusK   types.Nexus
+	coinType types.CoinType
+}
+
+func newCoin(ctx sdk.Context, k Keeper, nexusK types.Nexus, c sdk.Coin) coin {
+	return coin{
+		Coin:     c,
+		ctx:      ctx,
+		k:        k,
+		nexusK:   nexusK,
+		coinType: getCoinType(ctx, nexusK, c.Denom),
+	}
+}
+
+// lock locks coin from deposit address to escrow address
+func (c coin) lock(ibcK IBCKeeper, bankK types.BankKeeper, depositAddr sdk.AccAddress) error {
+	switch c.coinType {
+	case types.ICS20:
+		// get base denomination and tracing path
+		denomTrace, err := ibcK.ParseIBCDenom(c.ctx, c.Denom)
+		if err != nil {
+			return err
+		}
+
+		err = c.validateDenomTrace(denomTrace)
+		if err != nil {
+			return err
+		}
+
+		// lock tokens in escrow address
+		escrowAddress := types.GetEscrowAddress(c.Denom)
+		if err := bankK.SendCoins(
+			c.ctx, depositAddr, escrowAddress, sdk.NewCoins(c.Coin),
+		); err != nil {
+			return err
+		}
+	case types.Native:
+		// lock tokens in escrow address
+		escrowAddress := types.GetEscrowAddress(c.GetDenom())
+		if err := bankK.SendCoins(
+			c.ctx, depositAddr, escrowAddress, sdk.NewCoins(c.Coin),
+		); err != nil {
+			return err
+		}
+	case types.External:
+		// transfer coins from linked address to module account and burn them
+		if err := bankK.SendCoinsFromAccountToModule(
+			c.ctx, depositAddr, types.ModuleName, sdk.NewCoins(c.Coin),
+		); err != nil {
+			return err
+		}
+
+		// NOTE: should not happen as the module account was
+		// retrieved on the step above, and it has enough balance
+		// to burn.
+		funcs.MustNoErr(bankK.BurnCoins(c.ctx, types.ModuleName, sdk.NewCoins(c.Coin)))
+	default:
+		return fmt.Errorf("unrecognized coin type %d", c.coinType)
+	}
+
+	return nil
+}
+
+func (c coin) normalizeDenom(ibcK IBCKeeper) (sdk.Coin, error) {
+	switch c.coinType {
+	case types.ICS20:
+		// get base denomination and tracing path
+		denomTrace, err := ibcK.ParseIBCDenom(c.ctx, c.Denom)
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+
+		// convert denomination from 'ibc/{hash}' to native asset that recognized by nexus module
+		return sdk.NewCoin(denomTrace.GetBaseDenom(), c.Amount), nil
+	default:
+		return c.Coin, nil
+	}
+}
```

Comment:
this feels like something that should be done during construction?

Reply:
moved to construction